### ISSUE TITLE
Issue with Gradient computations

### ISSUE THREAD

**Issue by nazanint** [pytorch/pytorch#31957]
Hi, 

We're hitting nans when computing the second gradient. Is there a way for us to figure out what Nans similar to the example below are caused by? Is it underflow or overflow or some other useful information?

val = tensor([8.1929e-01, 8.6647e-02, 9.4059e-02, 2.5957e-15, 1.1755e-38,
          1.1755e-38, 1.1755e-38, 1.1755e-38, 1.1755e-38, 1.1755e-38])

val.requires_grad_(True)
score = dist.Dirichlet(tensor([0.01] * 10)).log_prob(val).sum() + dist.Categorical(val).log_prob(out).sum()

from torch.autograd import grad

first_grad, = grad(score, val, create_graph=True)
print(first_grad)
for i in range(10):
    second_grad, = grad(first_grad[i], val, create_graph=True)
    print(second_grad)

And here is what we get:

`tensor([-2.4757e+01,  1.4268e+02,  5.2633e+01, -3.8140e+14, -8.4219e+37,
        -8.4219e+37, -8.4219e+37, -8.4219e+37, -8.4219e+37, -8.4219e+37],
       grad_fn=<AddBackward0>)
tensor([-190.3480, 1000.0082, 1000.0081, 1000.0082,       nan,       nan,
              nan,       nan,       nan,       nan], grad_fn=<AddBackward0>)
tensor([  1000.0082, -12187.7812,   1000.0082,   1000.0083,         nan,
                nan,         nan,         nan,         nan,         nan],
       grad_fn=<AddBackward0>)
tensor([  1000.0082,   1000.0082, -10191.2354,   1000.0082,         nan,
                nan,         nan,         nan,         nan,         nan],
       grad_fn=<AddBackward0>)
tensor([1.0000e+03, 1.0000e+03, 1.0000e+03, 1.4694e+29,        nan,        nan,
               nan,        nan,        nan,        nan],
       grad_fn=<AddBackward0>)
tensor([1000.0083, 1000.0084, 1000.0082, 1000.0084,       inf,       nan,
              nan,       nan,       nan,       nan], grad_fn=<AddBackward0>)
tensor([1000.0083, 1000.0084, 1000.0082, 1000.0084,       nan,       inf,
              nan,       nan,       nan,       nan], grad_fn=<AddBackward0>)
tensor([1000.0083, 1000.0084, 1000.0082, 1000.0084,       nan,       nan,
              inf,       nan,       nan,       nan], grad_fn=<AddBackward0>)
tensor([1000.0083, 1000.0084, 1000.0082, 1000.0084,       nan,       nan,
              nan,       inf,       nan,       nan], grad_fn=<AddBackward0>)
tensor([1000.0083, 1000.0084, 1000.0082, 1000.0084,       nan,       nan,
              nan,       nan,       inf,       nan], grad_fn=<AddBackward0>)
tensor([1000.0083, 1000.0084, 1000.0082, 1000.0084,       nan,       nan,
              nan,       nan,       nan,       inf], grad_fn=<AddBackward0>)
 `

cc @ezyang @SsnL @albanD @zou3519 @gqchen

**Comment by zou3519** [comment]
You can use autograd's detect_anomaly mode might help in investigating nans https://pytorch.org/docs/stable/autograd.html#torch.autograd.detect_anomaly .

**Comment by albanD** [comment]
Closing this as there is no progress.
Please post on the forum if your have further questions: https://discuss.pytorch.org/
